include/seastar/util/chunked_deque.hh

```diff
@@ -0,0 +1,733 @@
+};
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline
+chunked_deque<T, items_per_chunk, save_free_chunks>::chunked_deque(chunked_deque&& x) noexcept
+        : _front_chunk(std::exchange(x._front_chunk, nullptr))
+        , _back_chunk(std::exchange(x._back_chunk, nullptr))
+        , _nchunks(std::exchange(x._nchunks, 0))
+        , _free_chunks(std::exchange(x._free_chunks, nullptr))
+        , _nfree_chunks(std::exchange(x._nfree_chunks, 0))
+{}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline
+chunked_deque<T, items_per_chunk, save_free_chunks>&
+chunked_deque<T, items_per_chunk, save_free_chunks>::operator=(chunked_deque&& x) noexcept {
+    if (&x != this) {
+        this->~chunked_deque();
+        new (this) chunked_deque(std::move(x));
+    }
+    return *this;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline bool
+chunked_deque<T, items_per_chunk, save_free_chunks>::empty() const noexcept {
+    return _front_chunk == nullptr;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline size_t
+chunked_deque<T, items_per_chunk, save_free_chunks>::size() const noexcept{
+    if (_front_chunk == nullptr) {
+        return 0;
+    } else if (_back_chunk == _front_chunk) {
+        // Single chunk.
+        return _front_chunk->size();
+    } else {
+        return _front_chunk->size() + _back_chunk->size()
+                + (_nchunks - 2) * items_per_chunk;
+    }
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+void chunked_deque<T, items_per_chunk, save_free_chunks>::clear() noexcept {
+    while (!empty()) {
+        pop_front();
+    }
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks> void
+chunked_deque<T, items_per_chunk, save_free_chunks>::shrink_to_fit() noexcept {
+    while (_free_chunks) {
+        auto next = _free_chunks->next;
+        delete _free_chunks;
+        _free_chunks = next;
+    }
+    _nfree_chunks = 0;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+chunked_deque<T, items_per_chunk, save_free_chunks>::~chunked_deque() {
+    clear();
+    shrink_to_fit();
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+typename chunked_deque<T, items_per_chunk, save_free_chunks>::chunk*
+chunked_deque<T, items_per_chunk, save_free_chunks>::new_chunk() {
+    chunk* ret;
+    if (_free_chunks) {
+        ret = _free_chunks;
+        _free_chunks = _free_chunks->next;
+        --_nfree_chunks;
+    } else {
+        ret = new chunk;
+    }
+    _nchunks++;
+    return ret;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::ensure_room_front() {
+    // If we don't have a back chunk or it's full, we need to create a new one
+    if (_front_chunk != nullptr && _front_chunk->begin > 0) {
+        return;
```

Comment:
I don't understand why _front_chunk->begin > 0 means the front chunk is full... Doesn't it just mean something was pushed into it? I.e., that it's not empty (but it's not full..)

I probably don't understand how you use the chunks - both the code here and the items_per_chunk/2 below I simply don't understand... In chunked_queue, a single chunk is a circular buffer - if you push and pop items and never go over 128 items, you're always using the same chunk. Is this also the case in this code, or you went for something simpler? I just failed to understand.

Reply:
Like you noticed below, this implementation doesn't use the chunks as circular buffers (and so need no calls to `mask()`.

Maybe that was a mistake since it may lead to under-utilizing the first chunk.

Maybe a better approach would to refactor the chinked_fifo code and use a common circular chunk class that will have both next and prev pointers and can be pushed to and popped from on either sides.

Reply:
I'll also look into how similar it is with a circular_buffer, and if it can shared there too.

Reply:
Maybe not making the chunks circular makes sense, if you believe it improves performance (e.g., maybe slightly faster indexing? or slightly simpler chunk-is-full condition?). But I don't know if it does

Reply:
I believe that it may improve performance slightly since there is no need to mask the indices (there is also no need to have the number of items in a chunk a power of 2, which can help plan chunks better aligned).

But the reason I chose to do do was to simplify the logic, especially around negative values of begin and end, that would be possible in a deque when items may be pushed to front and popped from back.

The bidirectional iterators use item_index values that are negative or greater than or equal to items_in_chunk as special values so they'll have to also check that the _chunk pointer is null to see if they are special or not, or to add another member to mark the iterator as pre-front or post-back.  That's doable...